### PR TITLE
feature: support torch_distributed distribution for Trainium instances

### PR DESCRIPTION
**Pull request by satishpasumarthi**
*Issue #, if available:*

*Description of changes:*
This PR is to add support for new distribution strategy for AWS trainium instances for Training in SageMaker.  Below are the related PRs for this effort:
- SM Toolkit PR: https://github.com/aws/sagemaker-training-toolkit/pull/154
- PT Toolkit PR: https://github.com/aws/sagemaker-pytorch-training-toolkit/pull/248

*Testing done:*
- Added unit and integration tests
- Added validation checks for supported distributions with Trainium

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 1]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
+<https://pytorch.org/docs/master/distributed.html#torch.distributed.init_process_group>`_
+with the ``xla`` backend as shown below.
+
+.. code:: python
+
+    import torch.distributed as dist
+
+    dist.init_process_group('xla')
+
+SageMaker takes care of ``'MASTER_ADDR'`` and ``'MASTER_PORT'`` for you via `torchrun`
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:326]:
```suggestion
SageMaker takes care of ``'MASTER_ADDR'`` and ``'MASTER_PORT'`` for you via ``torchrun``.
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 2]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
+
+For detailed documentation about modifying your training script for Trainium. Please refer `AWS Neuron Documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/mlp.html?highlight=torchrun#id5>`_
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:328]:
```suggestion
For detailed documentation about modifying your training script for Trainium, see `Multi-worker data-parallel MLP training using torchrun <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/mlp.html?highlight=torchrun#multi-worker-data-parallel-mlp-training-using-torchrun>`_ in the *AWS Neuron Documentation*.
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 3]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
 
+SageMaker Training on Trainium instances now supports the `xla`
+package through `torchrun`. With this, you do not need to manually pass RANK, 
+WORLD_SIZE, MASTER_ADDR, and MASTER_PORT. You can launch the training job using the
+:class:`sagemaker.pytorch.estimator.PyTorch` estimator class
+with the ``torch_distributed`` option as the distribution strategy.
+
+.. note::
+
+  This torch_distributed support is available
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:306]:
```suggestion
  This ``torch_distributed`` support is available
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 4]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
+  in the SageMaker Trainium (trn1) PyTorch Deep Learning Containers starting v1.11.0
+  `Supported Containers <https://github.com/aws/deep-learning-containers/blob/master/available_images.md#neuron-containers>`_
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:308]:
```suggestion
  To find a complete list of supported versions of PyTorch Neuron, see `Neuron Containers <https://github.com/aws/deep-learning-containers/blob/master/available_images.md#neuron-containers>`_ in the *AWS Deep Learning Containers GitHub repository*.
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 5]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
+SageMaker Training on Trainium instances now supports the `xla`
+package through `torchrun`. With this, you do not need to manually pass RANK, 
+WORLD_SIZE, MASTER_ADDR, and MASTER_PORT. You can launch the training job using the
+:class:`sagemaker.pytorch.estimator.PyTorch` estimator class
+with the ``torch_distributed`` option as the distribution strategy.
+
+.. note::
+
+  This torch_distributed support is available
+  in the SageMaker Trainium (trn1) PyTorch Deep Learning Containers starting v1.11.0
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:307]:
```suggestion
  in the SageMaker Trainium (trn1) PyTorch Deep Learning Containers starting v1.11.0.
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 6]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
+
+SageMaker takes care of ``'MASTER_ADDR'`` and ``'MASTER_PORT'`` for you via `torchrun`
+
+For detailed documentation about modifying your training script for Trainium. Please refer `AWS Neuron Documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/mlp.html?highlight=torchrun#id5>`_
+
+**Currently Supported backends:**
+
+-  ``xla`` for Trainium (Trn1) instances
+
+For up-to-date information on supported backends for Trainium instances, please refer `AWS Neuron Documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/index.html>`_
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:334]:
```suggestion
For up-to-date information on supported backends for Trainium instances, see `AWS Neuron Documentation <https://awsdocs-neuron.readthedocs-hosted.com/en/latest/index.html>`_.
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 7]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
+------------------------------------
+
+You can run multi-node distributed PyTorch training jobs on Trainium instances using the
+:class:`sagemaker.pytorch.estimator.PyTorch` estimator class.
+With ``instance_count=1``, the estimator submits a
+single-node training job to SageMaker; with ``instance_count`` greater
+than one, a multi-node training job is launched.
+
+With the ``torch_distributed`` option, the SageMaker PyTorch estimator runs a SageMaker
+training container for PyTorch, sets up the environment, and launches
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:346]:
```suggestion
training container for PyTorch Neuron, sets up the environment, and launches
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 8]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
...
+the training job using the ``torchrun`` command on each worker with the given information
+during the PyTorch DDP initialization.
+
+.. note::
+
+The following example shows how to run a PyTorch training using `torch_distributed` in SageMaker
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:352]:
```suggestion
The following example shows how to run a PyTorch training using ``torch_distributed`` in SageMaker
```

**Review by mchoi8739 — CHANGES_REQUESTED** [review]
Thank you so much for putting the new doc section. I made some minor suggestions. Please accept them as a batch if those looks good to you.

File: src/sagemaker/pytorch/estimator.py [hunk 9]
```diff
@@ -167,6 +168,18 @@ def __init__(
...
+
+                    .. code:: python
+                    {
+                        "torch_distributed": {
+                            "enabled": True
+                        }
+                    }
+                    To learn more, see `Distributed PyTorch Training on Trainium 
+                    <https://sagemaker.readthedocs.io/en/stable/frameworks/pytorch/using_pytorch.html#distributed-pytorch-training-on-trainium>`_.
+                `
```
Comment (mchoi8739) [inline src/sagemaker/pytorch/estimator.py:182]:
```suggestion
```

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 10]
```diff
@@ -292,7 +292,105 @@ using two ``ml.p4d.24xlarge`` instances:
 
     pt_estimator.fit("s3://bucket/path/to/training/data")
 
+Distributed PyTorch Training on Trainum (trn1) Instances
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:295]:
```suggestion
.. _distributed-pytorch-training-on-trainium:

Distributed PyTorch Training on Trainum (trn1) Instances
```

Comment (mchoi8739) [reply doc/frameworks/pytorch/using_pytorch.rst:295]: This is to make sure the link you added in `estimator.py` works.

File: src/sagemaker/fw_utils.py [hunk 11]
```diff
@@ -134,6 +134,14 @@
     "1.12.0",
 ]
 
+TORCH_DISTRIBUTED_SUPPORTED_FRAMEWORK_VERSIONS = [
+    "1.11.0"
```
Comment (vishwakaria) [inline src/sagemaker/fw_utils.py:138]: The documentation at line 307 says `torch-distributed` is available in the SageMaker Trainium (trn1) PyTorch Deep Learning Containers starting v1.11.0. But the validation seems to allow only 1.11 as the supported framework.
Let's make sure the two of them are in sync.

**Review by vishwakaria — CHANGES_REQUESTED** [review]

File: src/sagemaker/fw_utils.py [hunk 12]
```diff
@@ -808,11 +832,52 @@ def validate_distribution(
...
+    if match and match[1].startswith("trn"):
+        keys = distribution.keys()
+        if len(keys) == 0:
+            return
+        elif len(keys) == 1:
+            distribution_strategy = keys[0]
+            if distribution_strategy != "torch_distributed":
+                err_msg += (
+                f"Provided distribution strategy {distribution_strategy} is not supported by"
+                " Trainium instances yet.\n"
```
Comment (josephevans) [inline src/sagemaker/fw_utils.py:868]: Unless we guarantee to support all, I think we should remove the "yet" from this error message (and one below.)

Comment (mchoi8739) [reply src/sagemaker/fw_utils.py:868]: +1

File: src/sagemaker/fw_utils.py [hunk 13]
```diff
@@ -870,6 +935,73 @@ def validate_pytorch_distribution(
...
+    """
+    if framework_name and framework_name != "pytorch":
+        # We need to validate only for PyTorch framework
+        return
+
+    torch_distributed_enabled = False
+    if "torch_distributed" in distribution:
+        torch_distributed_enabled = distribution.get("torch_distributed").get("enabled", False)
+    if not torch_distributed_enabled:
+        # Distribution strategy other than pytorchddp is selected
```
[the remark this answers is not in the document]
Comment (satishpasumarthi) [reply src/sagemaker/fw_utils.py]: Done

File: src/sagemaker/fw_utils.py [hunk 14]
```diff
@@ -870,6 +935,73 @@ def validate_pytorch_distribution(
...
+                }
+        framework_name (str): A string representing the name of framework selected.
+        framework_version (str): A string representing the framework version selected.
+        py_version (str): A string representing the python version selected.
+        image_uri (str): A string representing a Docker image URI.
+
+    Raises:
+        ValueError: if
+            `py_version` is not python3 or
+            `framework_version` is not in TORCH_DISTRIBUTED_SUPPORTED_FRAMEWORK_VERSIONS
```
[the remark this answers is not in the document]
Comment (satishpasumarthi) [reply src/sagemaker/fw_utils.py:970]: Thanks for pointing this out. Updated the function name

File: src/sagemaker/fw_utils.py (hunk 11 above)
Comment (satishpasumarthi) [reply src/sagemaker/fw_utils.py:139]: Updated

File: src/sagemaker/fw_utils.py (hunk 12 above)
Comment (satishpasumarthi) [reply src/sagemaker/fw_utils.py]: Updated the wording.

File: doc/frameworks/pytorch/using_pytorch.rst [hunk 15]
```diff
@@ -292,7 +292,104 @@ using two ``ml.p4d.24xlarge`` instances:
 
     pt_estimator.fit("s3://bucket/path/to/training/data")
 
+.. _distributed-pytorch-training-on-trainium:
+=============================================
```
Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:296]: Need to bring the title back. `.. _distributed-pytorch-training-on-trainium:` is only a flag that enables links.

Comment (mchoi8739) [inline doc/frameworks/pytorch/using_pytorch.rst:296]:
```suggestion

Distributed PyTorch Training on Trainium
========================================
```

**Review by mchoi8739 — CHANGES_REQUESTED** [review]

File: doc/frameworks/pytorch/using_pytorch.rst (hunk 15 above)
Comment (satishpasumarthi) [reply doc/frameworks/pytorch/using_pytorch.rst]: Reverted this change to earlier version. Can you check now?

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run slow-tests

**Review by vishwakaria — APPROVED** [review]

**Review by mufaddal-rohawala — APPROVED** [review]

**Review by mchoi8739 — APPROVED** [review]
Thank you for incorporating doc feedback

**Review by knikure — APPROVED** [review]

**Pull request merged**